src/libs/SearchAutocompleteUtils.ts

```diff
@@ -0,0 +1,86 @@
+
+function parseForAutocomplete(text: string) {
+    try {
+        const parsedAutocomplete = autocompleteParser.parse(text) as SearchAutocompleteResult;
+        return parsedAutocomplete;
+    } catch (e) {
+        console.error(`Error when parsing autocopmlete query}"`, e);
+    }
+}
+
+function getAutocompleteTags(allPoliciesTagsLists: OnyxCollection<PolicyTagLists>, policyID?: string) {
+    const singlePolicyTagsList: PolicyTagLists | undefined = allPoliciesTagsLists?.[`${ONYXKEYS.COLLECTION.POLICY_TAGS}${policyID}`];
+    if (!singlePolicyTagsList) {
+        const uniqueTagNames = new Set<string>();
+        const tagListsUnpacked = Object.values(allPoliciesTagsLists ?? {}).filter((item) => !!item) as PolicyTagLists[];
+        tagListsUnpacked
+            .map(getTagNamesFromTagsLists)
+            .flat()
+            .forEach((tag) => uniqueTagNames.add(tag));
+        return Array.from(uniqueTagNames);
+    }
+    return getTagNamesFromTagsLists(singlePolicyTagsList);
+}
+
+function getAutocompleteRecentTags(allRecentTags: OnyxCollection<RecentlyUsedTags>, policyID?: string) {
+    const singlePolicyRecentTags: RecentlyUsedTags | undefined = allRecentTags?.[`${ONYXKEYS.COLLECTION.POLICY_RECENTLY_USED_TAGS}${policyID}`];
+    if (!singlePolicyRecentTags) {
+        const uniqueTagNames = new Set<string>();
+        Object.values(allRecentTags ?? {})
+            .map((testVar) => Object.values(testVar ?? {}))
+            .flat(2)
+            .forEach((tag) => uniqueTagNames.add(tag));
+        return Array.from(uniqueTagNames);
+    }
+    return Object.values(singlePolicyRecentTags ?? {}).flat(2);
+}
+
+function getAutocompleteCategories(allPolicyCategories: OnyxCollection<PolicyCategories>, policyID?: string) {
+    const singlePolicyCategories = allPolicyCategories?.[`${ONYXKEYS.COLLECTION.POLICY_CATEGORIES}${policyID}`];
+    if (!singlePolicyCategories) {
+        const uniqueCategoryNames = new Set<string>();
+        Object.values(allPolicyCategories ?? {}).map((policyCategories) => Object.values(policyCategories ?? {}).forEach((category) => uniqueCategoryNames.add(category.name)));
+        return Array.from(uniqueCategoryNames);
+    }
+    return Object.values(singlePolicyCategories ?? {}).map((category) => category.name);
+}
+
+function getAutocompleteRecentCategories(allRecentCategories: OnyxCollection<RecentlyUsedCategories>, policyID?: string) {
+    const singlePolicyRecentCategories = allRecentCategories?.[`${ONYXKEYS.COLLECTION.POLICY_RECENTLY_USED_CATEGORIES}${policyID}`];
+    if (!singlePolicyRecentCategories) {
+        const uniqueCategoryNames = new Set<string>();
+        Object.values(allRecentCategories ?? {}).map((policyCategories) => Object.values(policyCategories ?? {}).forEach((category) => uniqueCategoryNames.add(category)));
+        return Array.from(uniqueCategoryNames);
+    }
+    return Object.values(singlePolicyRecentCategories ?? {}).map((category) => category);
+}
```

Comment:
NAB maybe we could DRY the logic in some of these functions too since the logic is 95% identical

Reply:
Same as in https://github.com/Expensify/App/pull/51237#discussion_r1819798963